Adds the user to volunteer group

sahil9001 opened the pull request:
### Fixes #64 

### Description
Added the functionality to add the user to particular 'Volunteer' group of the organisation with a certain [org-id]. Not added tests as was having trouble writing it. 

**Type of Change:** (Delete irrelevant options)
- Code
- User Interface
- Documentation

**Code/Quality Assurance Only** (Delete irrelevant options)

- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)


### How Has This Been Tested?
Tested locally without code, adds the user to the group.


### Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged
- [x] Update Swagger documentation and the exported file at /docs folder
- [ ] Update requirements.txt

**Code/Quality Assurance**
- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

sahil9001 commented:
@naveennvrgup sir I've added tests to the PR.

naveennvrgup commented on org/views.py:
```diff
@@ -51,4 +52,41 @@ def post(self,request):
+        )
+        return Response({'You are added as a volunteer'},status.HTTP_201_CREATED)
+    else:
+        return Response({'Organization not present'},status.HTTP_400_BAD_REQUEST)
```
Match the response according to swagger docs u wrote

expected:
```
{
    "detail": "Organization not present"
}
```

actual:
![image](https://user-images.githubusercontent.com/33046846/82287829-ba7fe000-99be-11ea-9995-74d80493dece.png)

naveennvrgup commented on org/views.py:
```diff
@@ -51,4 +52,41 @@ def post(self,request):
+    org_count = Org.objects.filter(pk=org_id).count()
+    if org_count>0:
+        org = Org.objects.get(pk=org_id)
+        volunteer_permissions = Permissions()
+        volunteer_permission_set = PermissionSet.objects.get(
+            name='Volunteer',
+            org=org,
+            permissions = volunteer_permissions,
+        )
```
```suggestion
        volunteer_permission_set = PermissionSet.objects.get(
            name='Volunteer',
            org=org
        )
```
I think this line is redundant

naveennvrgup commented on org/views.py:
```diff
@@ -51,4 +52,41 @@ def post(self,request):
+            org=org,
+            permissions = volunteer_permissions,
+        )
+        volunteer_group = Group.objects.get(
+            name='Volunteer',
+            org=org,
+            default_permission_set=volunteer_permission_set,
+        )
```
```suggestion
        volunteer_group = Group.objects.get(
            name='Volunteer',
            org=org
        )
```

I think this line is redundant

naveennvrgup commented on org/views.py:
```diff
@@ -51,4 +52,41 @@ def post(self,request):
+            group = volunteer_group,
+            permissions = volunteer_permission_set 
+        )
+        return Response({'You are added as a volunteer'},status.HTTP_201_CREATED)
```
Match the response according to swagger docs u wrote

expected:
![image](https://user-images.githubusercontent.com/33046846/82288210-9c66af80-99bf-11ea-8334-bde0bbd5e7bd.png)

naveennvrgup commented on org/views.py:
```diff
@@ -51,4 +52,41 @@ def post(self,request):
+    operation_description="When an authenticated user hits this API it gets added to the volunteer group",
+    method='get',
+    responses={
+        '200': set_example({"message": "You're added as a volunteer!"}),
+        '401': set_example({"detail": "Authentication credentials were not provided."})
```
1. Please extract these responses to  `org.responses.py`
2. This response is not been documented 
![image](https://user-images.githubusercontent.com/33046846/82288352-e9e31c80-99bf-11ea-9d76-aacffb3e1c65.png)

naveennvrgup commented on tests/org/test_add_volunteer_api.py:
```diff
@@ -0,0 +1,90 @@
+        # Authorizing the requests by adding the token in header.
+        auth_client.credentials(HTTP_AUTHORIZATION=auth_token)
+        return auth_client
+
```
Great job with the testcases I would like you to add:
1. test_fail_bad_org_id
1. test_fail_already_member_of_org

sahil9001 commented:
@naveennvrgup I've modified the files as per you asked, please review it once again.

naveennvrgup commented on org/views.py:
```diff
@@ -51,4 +52,48 @@ def post(self,request):
+        '201': set_example(responses.add_volunteer_201),
+        '400': set_example(responses.org_not_present_400),
+        '401': set_example(responses.user_unauthorized_401),
+        '200': set_example(responses.user_already_present_200)
```
```suggestion
        '409': set_example(responses.user_already_present_200)
```
 I guess "409 conflict" will be more appropriate :thinking: . What u think?

naveennvrgup commented on tests/org/test_add_volunteer_api.py:
```diff
@@ -0,0 +1,123 @@
+            tagline='test',
+        )
+
+        # Create default group permissions  
```
A better way to do the setup will be to call the create org API or use the CreateOrgSerializer which has a custom save() has present.

sahil9001 replied:
I'm having trouble setting this can't we use the existing code for set up?

naveennvrgup replied:
I will provide you better explanation over call

naveennvrgup commented on tests/org/test_add_volunteer_api.py:
```diff
@@ -0,0 +1,123 @@
+            phone='5446465452'
+        )
+    
+    def create_auth_client(self):
```
By using `tests.AuthAPITestCase.AuthAPITestCase` we can avoid this code duplication. 

Ex. `tests.org. CreateOrgAPITestCase`

naveennvrgup commented on tests/org/test_add_volunteer_api.py:
```diff
@@ -0,0 +1,123 @@
+        return auth_client
+
+    #Creates auth client in the org as volunteer
+    def create_auth_client_in_org(self):
```
Code duplication in this function also. Try to use `AuthAPITestCase`

naveennvrgup commented on tests/AuthAPITestCase.py:
```diff
@@ -1,7 +1,8 @@
-
+from org.custom_model_field import PermissionSet as Permissions
+import uuid
 
```
1. I suppose you misunderstood the suggestions about `tests/AuthAPITestCase.py `.
1. The purpose is not to alter this file but to use it ( instead of duplicating the `create_auth_client()` in every test.
1. Please revoke the changes you made to this file :smiley:

sahil9001 commented:
@naveennvrgup sir I've made the following changes:

1. Removed duplicated code 
2. Modified test_already_member function to make request 2 times.
3. Used CreateOrgSerializer() to create org and inherited setUp() from AuthAPITestCase
